src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -167,21 +168,185 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
             {
                 try
                 {
+                    var utcNow = DateTimeOffset.UtcNow;
+                    IEnumerable<KeyValueWatcher> cacheExpiredWatchers = _options.ChangeWatchers.Where(changeWatcher => utcNow >= changeWatcher.CacheExpires);
+                    IEnumerable<KeyValueWatcher> cacheExpiredMultiKeyWatchers = _options.MultiKeyWatchers.Where(changeWatcher => utcNow >= changeWatcher.CacheExpires);
+
+                    // Skip refresh if the applicationSettings are loaded, but none of the watchers or adapters cache is expired.
+                    if (_applicationSettings != null &&
+                        !cacheExpiredWatchers.Any() &&
+                        !cacheExpiredMultiKeyWatchers.Any() &&
+                        !_options.Adapters.Any(adapter => adapter.NeedsRefresh()))
+                    {
+                        return;
+                    }
+
+                    IEnumerable<ConfigurationClient> availableClients = _configClientManager.GetAvailableClients();

```

Comment:
I had thought of this but didn't implement initially because then `GetAvailableClients` won't guarantee it will always check the clients available at that specific time (i.e., caller could pass a timestamp in future to get clients which are still in the backed-off state).

But now thinking of it, I think all the callers are private methods, we control who passes the timestamp, so it should be okay.

Reply:
Done!